tests/virt_autotest/kubevirt_tests_server.pm

```diff
@@ -0,0 +1,255 @@
+            } elsif ($key == 'GINKGO_SKIP') {
+                $skip_test = "-ginkgo.skip=\"${$section}{$key}\"";
+            } elsif ($key == 'EXTRA_OPT') {
+                $extra_opt = ${$section}{$key};
+            }
+        }
+        $artifacts = "/tmp/artifacts/$section";
+        $junit_xml = "/tmp/artifacts/$section\_test.xml";
+        $test_log = "/tmp/artifacts/$section\_test.log";
+        script_run("virt-tests -ginkgo.focus=\"$go_test\" \
+                    $skip_test \
+                    -ginkgo.slowSpecThreshold 60 \
+                    -kubeconfig=/root/.kube/config \
+                    -kubectl-path=`which kubectl` \
+                    -virtctl-path=`which virtctl` \
+                    -installed-namespace=kubevirt \
+                    -deploy-testing-infra=false \
+                    -config=/usr/share/kube-virt/manifests/testing/default-config.json \
+                    -ginkgo.v=true \
+                    -test.v=true \
+                    -ginkgo.trace=true \
+                    -ginkgo.noisySkippings=false \
+                    -ginkgo.progress=true \
+                    -ginkgo.noColor \
+                    $extra_opt \
+                    -artifacts=$artifacts \
+                    -junit-output=$junit_xml \
+                    -test.testlogfile=$test_log");
+		
+        assert_script_run("sed -ie 's/Tests Suite/$section/g' $junit_xml");
+    }
+}
+
+sub generate_test_report {
+    my $result_path = "/tmp/artifacts";
+    my $build_xml = "./buildTestReports.xml";
+
+    script_run("cat > $build_xml <<__END
+<project name=\"genTestReport\" default=\"gen\" basedir=\".\">
+    <description>
+        Generate the HTML report from JUnit XML files
+    </description>
+    <target name=\"gen\">
+        <property name=\"genReportDir\" location=\"$result_path\"/>
+        <delete dir=\"$result_path/html\"/>
+        <mkdir dir=\"$result_path/html\"/>
+        <junitreport todir=\"$result_path\">
+            <fileset dir=\"$result_path\">
+                <include name=\"*_test.xml\" />
+            </fileset>
+            <report format=\"frames\" todir=\"$result_path/html\" />
+        </junitreport>
+    </target>
+</project>
+__END");
+
+    # Generate JUnit HTML aggregate test reports by ant
+    script_run("ant -buildfile $build_xml");
+}
+
+sub upload_results {
+    # Compress test logs as a tarball and upload to openqa job
+    my $test_artifacts = "/tmp/test_artifacts.tar.gz";
+    assert_script_run("tar czvf $test_artifacts /tmp/artifacts");
+    upload_logs $test_artifacts;
+
+    my $html_dir = "/tmp/artifacts/html";
+    opendir(DIR, $html_dir) or die "Can't open the directory.";
+    my @html_files = readdir DIR;
+    foreach (@html_files) {
+        upload_logs $_;
+    }
+}
+
+sub post_fail_hook {
+    # Getting container status, uploading test reports & error logs for analysis
+    select_console 'log-console';
+    upload_results();
+}
+
+1;
```

Comment:
Generally speaking, I also have some other concerns:
1. Make your test stable enough, namely no failures on steps that are not supposed to fail if there is no environment issue.
2. Run tests against different scenarios to make sure your tests can handle there as expected, happy path and unhappy path.
3. Verify agent and server communicate mechanism, happy path and unhappy path.
4. Verify your report functionality.
5. Verify your logs collecting functionality.

Reply:
For communication between agent and server, I think it can be improved a little bit, for example, give the other end a NOT-OK singal if something goes wrong, so the other end can fail immediately.

Reply:
Please also check autoinst-log of you verification runs to make sure there is no "Use uninitialized variables" error.

Reply:
Agree with above.

Reply:
@nanzhg  How about this comment? Resolved or not?

Reply:
> For communication between agent and server, I think it can be improved a little bit, for example, give the other end a NOT-OK singal if something goes wrong, so the other end can fail immediately.

As tested in verification run, if any one of the parallel job failed, the pair job will run into fail immediately as well.

Reply:
> Please also check autoinst-log of you verification runs to make sure there is no "Use uninitialized variables" error.

Yes, there is no such error in autoinst-log.

Reply:
> @nanzhg How about this comment? Resolved or not?

It should be resolved.